### ISSUE TITLE
Undefined reference when compiling (Windows 10 - DevC++)

### ISSUE THREAD

**Issue by goncaloperes** [whoshuu/cpr#348]
**Environment:** Dev-C++ (5.11)
**OS:** Windows 10 (64-bit)

I am downloading files from a URL and save them to local directory in C++ using CPR.

As I want to iterate through different URLs, changing only specific values, I created a for loop and the URL is changing properly.

Now, to download the files, I added the following to the loop:
`auto r = cpr::Get(cpr::Url{"https://www.Example.com/YEAR_%Y/MONTH_%m/XLV/INT_PBC_EV_H_1_%d_%m_%Y_%d_%m_%Y.xls"});`

However, when compiling, I am getting the following error:
```
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xd): undefined reference to `__imp_curl_easy_cleanup'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x17): undefined reference to `__imp_curl_slist_free_all'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x21): undefined reference to `__imp_curl_formfree'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x72): undefined reference to `__imp_curl_easy_init'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x11f): undefined reference to `__imp_curl_slist_append'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x23f): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x252): undefined reference to `__imp_curl_slist_free_all'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2c5): undefined reference to `cpr::Timeout::Milliseconds() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2cf): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x305): undefined reference to `cpr::Timeout::Milliseconds() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x30f): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x343): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x376): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x391): undefined reference to `cpr::Authentication::GetAuthString() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x3d6): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x3f1): undefined reference to `cpr::Digest::GetAuthString() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x432): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x46c): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x4cc): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x523): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x552): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x598): more undefined references to `__imp_curl_easy_setopt' follow
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x5a7): undefined reference to `cpr::Cookies::GetEncoded() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x62c): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x68c): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x6e9): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x73a): undefined reference to `cpr::VerifySsl::operator bool() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x74d): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x757): undefined reference to `cpr::VerifySsl::operator bool() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x7e8): undefined reference to `cpr::Timeout::Milliseconds() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x7f2): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x838): undefined reference to `cpr::Timeout::Milliseconds() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x842): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x889): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x8a4): undefined reference to `cpr::Authentication::GetAuthString() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x8e9): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x904): undefined reference to `cpr::Digest::GetAuthString() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x945): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x97f): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x9df): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xa36): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xa65): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xaab): more undefined references to `__imp_curl_easy_setopt' follow
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xaba): undefined reference to `cpr::Cookies::GetEncoded() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xb4f): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xbaf): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xc0c): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xc6d): undefined reference to `cpr::VerifySsl::operator bool() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xc80): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xc8a): undefined reference to `cpr::VerifySsl::operator bool() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xd18): undefined reference to `cpr::Timeout::Milliseconds() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xd22): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xd68): undefined reference to `cpr::Timeout::Milliseconds() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xd72): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xdb9): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xdd4): undefined reference to `cpr::Authentication::GetAuthString() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xe19): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xe34): undefined reference to `cpr::Digest::GetAuthString() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xe75): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xeaf): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xf0f): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xf66): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xf95): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xfdb): more undefined references to `__imp_curl_easy_setopt' follow
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0xfea): undefined reference to `cpr::Cookies::GetEncoded() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x107f): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x10df): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x113c): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x119d): undefined reference to `cpr::VerifySsl::operator bool() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x11b0): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x11ba): undefined reference to `cpr::VerifySsl::operator bool() const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1206): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x12ca): undefined reference to `__imp_curl_easy_init'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x14a6): undefined reference to `__imp_curl_version_info'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1554): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1936): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1988): undefined reference to `cpr::Proxies::has(std::string const&) const'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x199b): undefined reference to `cpr::Proxies::operator[](std::string const&)'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x19ee): undefined reference to `cpr::util::writeFunction(void*, unsigned long long, unsigned long long, std::string*)'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1a22): undefined reference to `__imp_curl_easy_perform'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1a3a): undefined reference to `__imp_curl_easy_getinfo'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1aeb): undefined reference to `cpr::util::split(std::string const&, char)'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1b5c): undefined reference to `cpr::Cookies::operator[](std::string const&)'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1bfc): undefined reference to `__imp_curl_slist_free_all'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1c15): undefined reference to `cpr::util::parseHeader(std::string const&)'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1c2e): undefined reference to `cpr::Error::getErrorCodeForCurlError(int)'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x1ef1): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x21a8): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2218): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2278): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x22d8): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2338): more undefined references to `__imp_curl_easy_setopt' follow
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2924): undefined reference to `__imp_curl_formadd'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2b71): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2b7f): undefined reference to `__imp_curl_formfree'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2c84): undefined reference to `__imp_curl_formadd'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2ed1): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text+0x2edf): undefined reference to `__imp_curl_formfree'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text.startup+0x153): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.text.startup+0x168): undefined reference to `__imp_curl_easy_setopt'
G:\Users\ADMIN~1\AppData\Local\Temp\ccOBPd8t.o:gmp_v02.cpp:(.rdata$.refptr._ZN3cpr4util13writeFunctionEPvyyPSs[.refptr._ZN3cpr4util13writeFunctionEPvyyPSs]+0x0): undefined reference to `cpr::util::writeFunction(void*, unsigned long long, unsigned long long, std::string*)'
collect2.exe: error: ld returned 1 exit status

Compilation results...
--------
- Errors: 1
- Warnings: 0
- Compilation Time: 3.52s
```

In case it helps, this is my #includes list:
```
#include <cstring>
#include <ctime>
#include <iostream>
#include <winsock2.h>
#include <windows.h>
#include <include/cpr/cpr.h>
#include <cpr/session.cpp>
#include <include/curl/curl.h>
```

From what I found, it seems to be a Linker error, more specifically, the linker seems to not be able to find the curl library.

**Comment by jungletek** [comment]
You don't need to explicitly `#include <curl.h`>, CPR already includes it.

You can control whether CPR uses the "system's" version of Curl by manipulating the `USE_SYSTEM_CURL` variable.

For example, my root CMakeLists.txt has these entries for setting up CPR and its dependencies:
```
set(BUILD_CPR_TESTS "OFF")
set(USE_SYSTEM_CURL "ON")
set(CMAKE_USE_OPENSSL "ON")

set(OPENSSL_DIR "C:/msys64/mingw64")
set(OPENSSL_INCLUDE_DIR ${OPENSSL_DIR}/include)
set(OPENSSL_LIBRARY ${OPENSSL_DIR}/lib/libssl.a)

set(CURL_PATH "C:/msys64/mingw64")
set(CURL_INCLUDE_DIR ${CURL_PATH}/include)
set(CURL_LIBRARY ${CURL_PATH}/lib/libcurl.a)
find_package(CURL REQUIRED)

set(CPR_DIR "cpr-feature-Stream-MOD")
add_subdirectory(${CPR_DIR} ${CPR_DIR} EXCLUDE_FROM_ALL)
set(CPR_INCLUDE_DIR ${CPR_DIR}/include)
set(CPR_LIBRARY ${CPR_DIR}/cpr)
find_package(CPR REQUIRED)
```

**Comment by COM8** [comment]
It looks to me like you did not initialized the submodules. So no libcurl source files/headers.
Try `git submodule update --init --recursive`.
Like @jungletek mentioned you also can use your systems CURL.

**Comment by COM8** [comment]
Should be fixed by #383 .
Feel free to reopen if this is still a problem.
